File: UnityClient/PlanetServer/Src/Data/PsObject.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Reflection;

using Pathfinding.Serialization.JsonFx;

namespace PS.Data
{
    /// <summary>
    /// PsObject is the basis of sending data to and from the server.  It can hold an arbitrary number of key-value pairs of primitive data types, 
    /// Lists of primitive data types, PsArrays, and other PsObjects.
    /// </summary>
    public class PsObject
    {
        private Dictionary<string, PsDataWrapper> _content;

        /// <summary>
        /// Create a PsObject from a JSON string.
        /// </summary>
        /// <param name="json">JSON to create PsObject.</param>
        /// <returns>PsObject containing values from JSON.</returns>
        public static PsObject Create(string json)
        {
            object obj = JsonReader.Deserialize(json);

            if (obj is IDictionary)
            {
                return Create((Dictionary<string, object>)obj);
            }
            else
            {
                throw new Exception("JSON not in proper format");
            }
        }

        /// <summary>
        /// Create a PsObject from Dictionary of string, object pairs.
        /// </summary>
        /// <param name="dict">Key-value pairs of data.</param>
        /// <returns>PsObject containing values from Dictionary.</returns>
        public static PsObject Create(Dictionary<string, object> dict)
        {
            PsObject psobj = new PsObject();

            foreach (KeyValuePair<string, object> item in dict)
            {
                string key = item.Key;
                Dictionary<string, object> value = (Dictionary<string, object>)item.Value;

                int type = (int)value["t"];

                switch (type)
                {
                    case (int)Constants.PsType.Boolean:
                        psobj.SetBoolean(key, Convert.ToBoolean(value["v"]));
                        break;

                    case (int)Constants.PsType.String:
                        psobj.SetString(key, Convert.ToString(value["v"]));
                        break;

                    case (int)Constants.PsType.Integer:
                        psobj.SetInt(key, Convert.ToInt32(value["v"]));
                        break;

                    case (int)Constants.PsType.Long:
                        psobj.SetLong(key, Convert.ToInt64(value["v"]));
                        break;

                    case (int)Constants.PsType.Float:
                        psobj.SetFloat(key, Convert.ToSingle(value["v"]));
                        break;

                    case (int)Constants.PsType.PSObject:
                        psobj.SetPsObject(key, PsObject.Create((Dictionary<string, object>)value["v"]));
                        break;

                    case (int)Constants.PsType.PSArray:
                        psobj.SetPsArray(key, PsArray.Create((object[])value["v"]));
                        break;

                    case (int)Constants.PsType.Number:
                        psobj.SetFloat(key, Convert.ToSingle(value["v"]));
                        break;
                }
            }

            return psobj;
        }

        /// <summary>
        /// Initializes a new instance of the PsObject class.
        /// </summary>
        public PsObject()
        {
            _content = new Dictionary<string, PsDataWrapper>();
        }

        internal PsObject(Dictionary<string, PsDataWrapper> content)
        {
            _content = content;
        }

        /// <summary>
        /// Get if PsObject contains a key.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>true if the key exisits, false if it doesn't.</returns>
        public bool HasKey(string key)
        {
            return _content.ContainsKey(key);
        }

        /// <summary>
        /// Look for a int value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Int value of key.</returns>
        public int GetInt(string key)
        {
            return (int)_content[key].v;
        }

        /// <summary>
        /// Look for a (Flash)number value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Number value of key.</returns>
        public float GetNumber(string key)
        {
            return (float)_content[key].v;
        }

        /// <summary>
        /// Look for a string value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>String value of key.</returns>
        public string GetString(string key)
        {
            return (string)_content[key].v;
        }

        /// <summary>
        /// Look for a boolean value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Boolean value of key.</returns>
        public bool GetBoolean(string key)
        {
            return (bool)_content[key].v;
        }

        /// <summary>
        /// Look for a float value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Float value of key.</returns>
        public float GetFloat(string key)
        {
            return (float)_content[key].v;
        }

        /// <summary>
        /// Look for a long value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Long value of key.</returns>
        public float GetLong(string key)
        {
            return (long)_content[key].v;
        }

        /// <summary>
        /// Look for a PsObject value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>PsObject value of key.</returns>
        public PsObject GetPsObject(string key)
        {
            return (PsObject)_content[key].v;
        }

        /// <summary>
        /// Look for a PsArray value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>PsArray value of key.</returns>
        public PsArray GetPsArray(string key)
        {
            return (PsArray)_content[key].v;
        }

        /// <summary>
        /// Look for a boolean array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Boolean array value of key.</returns>
        public List<bool> GetBoolArray(string key)
        {            
            PsArray psa = GetPsArray(key);

            List<bool> list = new List<bool>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetBoolean(i));
   
            return list;
        }

        /// <summary>
        /// Look for a string array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>String array value of key.</returns>
        public List<string> GetStringArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<string> list = new List<string>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetString(i));

            return list;
        }

        /// <summary>
        /// Look for a int array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Int array value of key.</returns>
        public List<int> GetIntArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<int> list = new List<int>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetInt(i));

            return list;
        }

        /// <summary>
        /// Look for a long array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Long array value of key.</returns>
        public List<long> GetLongArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<long> list = new List<long>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetLong(i));

            return list;
        }

        /// <summary>
        /// Look for a float array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>float array value of key.</returns>
        public List<float> GetFloatArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<float> list = new List<float>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetFloat(i));

            return list;
        }

        /// <summary>
        /// Look for a double array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Double array value of key.</returns>
        public List<double> GetDoubleArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<double> list = new List<double>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetDouble(i));

            return list;
        }

        /// <summary>
        /// Look for a PsObject array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>PsObject array value of key.</returns>
        public List<PsObject> GetPsObjectArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<PsObject> list = new List<PsObject>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetPsObject(i));

            return list;
        }

        /// <summary>
        /// Look for a (Flash)number array value in the PsObject.
        /// </summary>
        /// <param name="key">Key to look for.</param>
        /// <returns>Number array value of key.</returns>
        public List<double> GetNumberArray(string key)
        {
            PsArray psa = GetPsArray(key);

            List<double> list = new List<double>();
            for (int i = 0; i < psa.Count; ++i)
                list.Add(psa.GetNumber(i));

            return list;
        }

        /// <summary>
        /// Set a int in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetInt(string key, int value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.Integer, value);
        }

        /// <summary>
        /// Set a number in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetNumber(string key, float value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.Number, value);
        }

        /// <summary>
        /// Set a string in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetString(string key, string value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.String, value);
        }

        /// <summary>
        /// Set a boolean in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetBoolean(string key, bool value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.Boolean, value);
        }

        /// <summary>
        /// Set a float in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetFloat(string key, float value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.Float, value);
        }

        /// <summary>
        /// Set a long in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetLong(string key, long value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.Long, value);
        }

        /// <summary>
        /// Set a PsObject in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetPsObject(string key, PsObject value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.PSObject, value);
        }

        /// <summary>
        /// Set a PsArray in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetPsArray(string key, PsArray value)
        {
            _content[key] = new PsDataWrapper((int)Constants.PsType.PSArray, value);
        }

        /// <summary>
        /// Set a boolean array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetBooleanArray(string key, List<bool> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddBoolean(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a string array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetStringArray(string key, List<string> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddString(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a int array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetIntArray(string key, List<int> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddInt(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a long array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetLongArray(string key, List<long> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddLong(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a float array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetFloatArray(string key, List<float> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddFloat(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a double array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetDoubleArray(string key, List<double> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddDouble(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a PsObject array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetPsObjectArray(string key, List<PsObject> value)
        {
            PsArray array = new PsArray();
            
            for (int i = 0; i < value.Count; ++i)
            {
                array.AddPsObject(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Set a (Flash)number array in the PsObject.
        /// </summary>
        /// <param name="key">Key for value.</param>
        /// <param name="value">Value to add.</param>
        public void SetNumberArray(string key, List<double> value)
        {
            PsArray array = new PsArray();

            for (int i = 0; i < value.Count; ++i)
            {
                array.AddNumber(value[i]);
            }

            SetPsArray(key, array);
        }

        /// <summary>
        /// Create a raw object for JSON serialization.
        /// </summary>
        /// <returns>Object from PsObject.</returns>
        public object ToObject()
        {
            Dictionary<string, object> ret = new Dictionary<string, object>();

            string[] keys = new string[_content.Keys.Count];
            _content.Keys.CopyTo(keys, 0);

            for (int i = 0; i < keys.Length; ++i)
            {
                string key = keys[i];
                PsDataWrapper data = _content[key];

                if (data.v is PsObject)
                {
                    Hashtable hash = new Hashtable(2);
                    hash.Add(Constants.TYPE_FLAG, (int)Constants.PsType.PSObject);
                    hash.Add(Constants.VALUE_FLAG, (data.v as PsObject).ToObject());

                    ret[key] = hash;
                }
                else if (data.v is PsArray)
                {
                    Hashtable hash = new Hashtable(2);
                    hash.Add(Constants.TYPE_FLAG, (int)Constants.PsType.PSArray);
                    hash.Add(Constants.VALUE_FLAG, (data.v as PsArray).ToArrayObject());

                    ret[key] = hash;
                }
                else
                    ret[key] = data;
            }

            return ret;
        }
    }
}

File: PlanetServer/src/planetserver/network/PsObject.java
package planetserver.network;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import net.sf.json.JSONArray;
import net.sf.json.JSONObject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PsObject
{
    private static final Logger logger = LoggerFactory.getLogger(PsObject.class);
    
    private HashMap<String, PsDataWrapper> _content;
    public int version;

    public PsObject()
    {
        _content = new HashMap<String, PsDataWrapper>();
    }

    public boolean hasKey(String key)
    {
        return _content.containsKey(key);
    }

    public int getInteger(String key)
    {
        return (Integer)_content.get(key).getV();
    }

    public long getLong(String key)
    {
        return (Long)_content.get(key).getV();
    }

    public String getString(String key)
    {
        return (String)_content.get(key).getV();
    }

    public boolean getBoolean(String key)
    {
        Boolean b = (Boolean)_content.get(key).getV();
        return b;
    }

    public float getFloat(String key)
    {
        return (Float)_content.get(key).getV();
    }

    public double getDouble(String key)
    {
        return (Double)_content.get(key).getV();
    }

    public PsArray getPsArray(String key)
    {
        return (PsArray)_content.get(key).getV();
    }

    public PsObject getPsObject(String key)
    {
        return (PsObject)_content.get(key).getV();
    }
    
    //number from as3, just returns a double
    public double getNumber(String key)
    {
        return (Double)_content.get(key).getV();
    }
    
    public List<Boolean> getBoolArray(String name)
    {
        List<Boolean> list = new ArrayList<Boolean>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((Boolean)value);
        }
        return list;
    }
    
    public List<String> getStringArray(String name)
    {
        List<String> list = new ArrayList<String>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((String)value);
        }
        return list;
    }

    public List<Integer> getIntegerArray(String name)
    {
        List<Integer> list = new ArrayList<Integer>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((Integer)value);
        }
        return list;
    }

    public List<Long> getLongArray(String name)
    {
        List<Long> list = new ArrayList<Long>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((Long)value);
        }
        return list;
    }

    public List<Float> getFloatArray(String name)
    {
        List<Float> list = new ArrayList<Float>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((Float)value);
        }
        return list;
    }

    public List<Double> getDoubleArray(String name)
    {
        List<Double> list = new ArrayList<Double>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((Double)value);
        }
        return list;
    }

    public List<PsObject> getPsObjectArray(String name)
    {
        List<PsObject> list = new ArrayList<PsObject>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((PsObject)value);
        }
        return list;
    }
    
    public List<Double> getNumberArray(String name)
    {
        List<Double> list = new ArrayList<Double>();
        for (PsDataWrapper data : getPsArray(name).content)
        {
            Object value = data.getV();
            list.add((Double)value);
        }
        return list;
    }

    public void setBoolean(String key, boolean value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_BOOLEAN));
    }
    
    public void setString(String key, String value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_STRING));
    }
    
    public void setInteger(String key, int value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_INTEGER));
    }

    public void setLong(String key, Long value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_LONG));
    }
    
    public void setFloat(String key, float value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_FLOAT));
    }

    public void setDouble(String key, double value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_DOUBLE));
    }

    public void setPsObject(String key, PsObject value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_PSOBJECT));
    }
    
    public void setPsArray(String key, PsArray value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_PSARRAY));
    }

    public void setNumber(String key, double value)
    {
        _content.put(key, new PsDataWrapper(value, PsType.TYPE_NUMBER));
    }

    public void setBoolArray(String string, List<Boolean> clctn)
    {
        PsArray psArray = new PsArray();
        for (Boolean element : clctn)
        {
            psArray.addBoolean(element);
        }
        setPsArray(string, psArray);
    }
    
    public void setStringArray(String string, List<String> clctn)
    {
        PsArray psArray = new PsArray();
        for (String element : clctn)
        {
            psArray.addString(element);
        }
        setPsArray(string, psArray);
    }

    public void setIntegerArray(String string, List<Integer> clctn)
    {
        PsArray psArray = new PsArray();
        for (Integer element : clctn)
        {
            psArray.addInteger(element);
        }
        setPsArray(string, psArray);
    }

    public void setLongArray(String string, List<Long> clctn)
    {
        PsArray psArray = new PsArray();
        for (Long element : clctn)
        {
            psArray.addLong(element);
        }
        setPsArray(string, psArray);
    }

    public void setFloatArray(String string, List<Float> clctn)
    {
        PsArray psArray = new PsArray();
        for (Float element : clctn)
        {
            psArray.addFloat(element);
        }
        setPsArray(string, psArray);
    }

    public void setDoubleArray(String string, List<Double> clctn)
    {
        PsArray psArray = new PsArray();
        for (Double element : clctn)
        {
            psArray.addDouble(element);
        }
        setPsArray(string, psArray);
    }
 
    public void setPsObjectArray(String string, List<PsObject> clctn)
    {
        PsArray psArray = new PsArray();
        for (PsObject element : clctn)
        {
            psArray.addPsObject(element);
        }
        setPsArray(string, psArray);
    }

    public String toJsonString()
    {
        return toJSONObject().toString();
    }

    /**
     * Takes in json object and applies properties to the ps object
     * @param jsonObject 
     */
    public void fromJsonObject(JSONObject jsonObject)
    {
        Iterator<?> keys = jsonObject.keys();
        while (keys.hasNext())
        {
            String key = (String) keys.next();
            JSONObject keyObj = (JSONObject) jsonObject.get(key);

            String typeID = "t";
            String valueID = "v";

            int type = keyObj.getInt((typeID));
            Object value = keyObj.get(valueID);

            switch (type)
            {
                case PsType.TYPE_BOOLEAN:
                    setBoolean(key, (Boolean) value);
                    break;

                case PsType.TYPE_STRING:
                    setString(key, (String) value);
                    break;

                case PsType.TYPE_DOUBLE:
                    setDouble(key, Double.valueOf(keyObj.getDouble(valueID)));
                    break;

                case PsType.TYPE_FLOAT:
                    setFloat(key, (Float) value);
                    break;

                case PsType.TYPE_INTEGER:
                    setInteger(key, (Integer) value);
                    break;

                case PsType.TYPE_LONG:
                    setLong(key, Long.valueOf(keyObj.getLong(valueID)));
                    break;

                case PsType.TYPE_NUMBER:
                    setNumber(key, Double.valueOf(keyObj.getDouble(valueID)));
                    break;

                case PsType.TYPE_PSARRAY:
                    PsArray newArray = new PsArray();
                    newArray.fromJsonObject((JSONArray) value);
                    setPsArray(key, newArray);
                    break;

                case PsType.TYPE_PSOBJECT:
                    PsObject newObject = new PsObject();
                    newObject.fromJsonObject((JSONObject) value);
                    setPsObject(key, newObject);
                    break;

                default:
                    throw new Error("Unsupported Type!");

            }
        }
    }

    public JSONObject toJSONObject()
    {
        JSONObject jObject = new JSONObject();

        for (String key : _content.keySet())
        {
            PsDataWrapper obj = (PsDataWrapper) (_content.get(key));

            if (obj.getV() instanceof PsObject)
            {
                JSONObject psObject = new JSONObject();
                psObject.put(PsType.TYPE_FLAG, PsType.TYPE_PSOBJECT);
                psObject.put(PsType.VALUE_FLAG, ((PsObject) obj.getV()).toJSONObject());
                jObject.put(key, psObject);
            }
            else if (obj.getV() instanceof PsArray)
            {
                JSONObject psArray = new JSONObject();
                psArray.put(PsType.TYPE_FLAG, PsType.TYPE_PSARRAY);
                psArray.put(PsType.VALUE_FLAG, ((PsArray) obj.getV()).toJSONObject());
                jObject.put(key, psArray);
            }
            else
            {
                jObject.put(key, obj);
            }
        }
        return jObject;
    }

    public Map<String, Object> toMap()
    {
        Map<String, Object> map = new HashMap<String, Object>();

        for (String key : _content.keySet())
        {
            Object obj = _content.get(key).getV();

            if (obj instanceof PsObject)
            {
                map.put(key, ((PsObject) obj).toMap());
            }
            else if (obj instanceof PsArray)
            {
                map.put(key, ((PsArray) obj).toObjectList());
            }
            else
            {
                map.put(key, obj);
            }
        }

        return map;
    }
}
